clients/src/main/java/org/apache/kafka/clients/consumer/internals/MetadataSnapshot.java

```diff
@@ -0,0 +1,111 @@
+ */
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.common.utils.Utils;
+
+import java.security.MessageDigest;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+
+/**
+ * Represents a point-in-time view of the topic metadata corresponding to a
+ * consumer's metadata.
+ *
+ * NOT thread-safe
+ */
+public class MetadataSnapshot {
+
+    private final MessageDigest digest;
+    private final Set<String> localSubscribedTopics;
+    private final Set<String> groupSubscribedTopics;
+    private final SortedMap<String, Integer> metadata; // topic -> # partitions
+
+    public MetadataSnapshot(
+            Set<String> localSubscribedTopics,
+            Set<String> groupSubscribedTopics,
+            SortedMap<String, Integer> metadata) {
+        this.localSubscribedTopics = localSubscribedTopics;
+        this.groupSubscribedTopics = groupSubscribedTopics;
+        this.metadata = metadata;
+
+        try {
+            digest = MessageDigest.getInstance("SHA-256");
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Get the number of partitions per topic
+     * @return Map from topic to number of partitions
+     */
+    public SortedMap<String, Integer> topicMetadata() {
+        return metadata;
+    }
+
+    /**
+     * Get the topics the local consumer has subscribed to
+     * @return A set of the local subscribed topics
+     */
+    public Set<String> localSubscibedTopics() {
+        return localSubscribedTopics;
+    }
+
+    /**
+     * Get the topics the group has collectively subscribed to
+     * @return A set of the group subscribed topics
+     */
+    public Set<String> groupSubscribedTopics() {
+        return groupSubscribedTopics;
+    }
+
+    public byte[] hash() {
+        digest.reset();
+        for (Map.Entry<String, Integer> topicEntry : metadata.entrySet()) {
+            String topic = topicEntry.getKey();
+            Integer partitions = topicEntry.getValue();
+            digest.update(Utils.utf8(topic));
+            digest.update(Utils.toArrayLE(partitions));
+        }
+        return digest.digest();
+    }
+
+    public boolean contains(Set<String> topics) {
+        return metadata.keySet().containsAll(topics);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        MetadataSnapshot that = (MetadataSnapshot) o;
+
+        if (localSubscribedTopics != null ? !localSubscribedTopics.equals(that.localSubscribedTopics) : that.localSubscribedTopics != null)
+            return false;
+        if (groupSubscribedTopics != null ? !groupSubscribedTopics.equals(that.groupSubscribedTopics) : that.groupSubscribedTopics != null)
+            return false;
+        return !(metadata != null ? !metadata.equals(that.metadata) : that.metadata != null);
```

Comment:
`return metadata != null ? metadata.equals(that.metadata) : that.metadata == null;`